### PR TITLE
Support inode writecount

### PR DESCRIPTION
**Pull request by tanjianfeng**
This patchset is mainly to fix issue #1005.

### CODE REVIEW

File: pkg/sentry/fs/ramfs/dir.go [hunk 1]
```diff
@@ -327,7 +327,15 @@ func (d *Dir) Create(ctx context.Context, dir *fs.Inode, name string, flags fs.F
 	// Create the Dirent and corresponding file.
 	created := fs.NewDirent(ctx, inode, name)
 	defer created.DecRef()
-	return created.Inode.GetFile(ctx, created, flags)
+
+	f, err := created.Inode.GetFile(ctx, created, flags)
+	// Get an extra reference to pin inode in kernel. See Linux's
+	// fs/ramfd/inode.c:ramfs_mknod(). But here it also increments
```
[the remark this answers is not in the document]
Comment (nlacasse) [reply pkg/sentry/fs/ramfs/dir.go]: I don't see the call to get_write_access or put_write_access in that method, or anywhere in that file.

File: pkg/sentry/memmap/memmap.go [hunk 2]
```diff
@@ -254,6 +254,9 @@ type MappingIdentity interface {
 	// exclusive, hence mr.End-1.) It is defined rather than Fsync so that
 	// implementors don't need to depend on the fs package for fs.SyncType.
 	Msync(ctx context.Context, mr MappableRange) error
+
+	// DenyWrite is used to inc/dec inode's counter, WriteCount.
```
Comment (nlacasse) [inline pkg/sentry/memmap/memmap.go]: Instead of one method that takes a bool, why not two methods?  `DenyWrite()` and `AllowWrite()` ?  That makes the code much easier to read.

File: pkg/sentry/memmap/memmap.go [hunk 3]
```diff
@@ -342,6 +345,9 @@ type MMapOpts struct {
 	// that is exclusive to the MemoryManager.
 	Private bool
 
+	// DenyWrite is true if ...
```
Comment (nlacasse) [inline pkg/sentry/memmap/memmap.go]: ...

File: pkg/syserror/syserror.go [hunk 4]
```diff
@@ -63,6 +63,7 @@ var (
 	ENOTSOCK     = error(syscall.ENOTSOCK)
 	ENOTSUP      = error(syscall.ENOTSUP)
 	ENOTTY       = error(syscall.ENOTTY)
+	ETXTBSY      = error(syscall.ETXTBSY)
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/syserror/syserror.go]: Oh, it's in alphbetical order, instead of the order of macro definition. Will change. Thanks!

File: pkg/sentry/fs/file.go [hunk 5]
```diff
@@ -148,9 +148,14 @@ func (f *File) DecRef() {
...
 		f.Dirent.Inode.LockCtx.BSD.UnlockRegion(lock.UniqueID(f.UniqueID), lockRng)
 
+		if f.flags.Write {
+			f.Dirent.Inode.PutWriteAccess()
+		}
+
 		// Release resources held by the FileOperations.
 		f.FileOperations.Release()
 
+
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/file.go]: Thanks, will fix this in next version.

File: pkg/sentry/fs/inode.go [hunk 6]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+
+	// WriteCount can have the following values:
+	// - 0: no writers, no VM_DENYWRITE mappings
+	// - < 0: (-InodeWriteCount) vmas with VM_DENYWRITE set exist
+	// - > 0: (InodeWriteCount) users are writing to the file.
+	// Refer to Linux's i_writecount field of struct inode.
+	WriteCount int32
+}
+
+func atomicIncUnlessNegative(addr *int32) bool {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Make sense.

File: pkg/sentry/fs/inode.go [hunk 7]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+// Refer to Linux/include/linux/fs.h for below helper functions on WriteCount.
+
+func (i *Inode) GetWriteAccess() error {
+	if !IsRegular(i.StableAttr) {
+		return nil
+	}
+
+	if atomicIncUnlessNegative(&i.WriteCount) {
+		return nil
+	} else {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Will fix in next version.

File: pkg/sentry/fs/inode.go [hunk 8]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+		return
+	}
+
+	atomic.AddInt32(&i.WriteCount, -1)
+}
+
+func (i *Inode) DenyWriteAccess() error {
+	if atomicDecUnlessPositive(&i.WriteCount) {
+		return nil
+	} else {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 9]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+		if (atomic.CompareAndSwapInt32(addr, prev, next)) {
+			break;
+		}
+	}
+	return true
+}
+
+// Refer to Linux/include/linux/fs.h for below helper functions on WriteCount.
+
+func (i *Inode) GetWriteAccess() error {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 10]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+	}
+
+	if atomicIncUnlessNegative(&i.WriteCount) {
+		return nil
+	} else {
+		return syserror.ETXTBSY
+	}
+}
+
+func (i *Inode) PutWriteAccess() {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 11]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+	if !IsRegular(i.StableAttr) {
+		return
+	}
+
+	atomic.AddInt32(&i.WriteCount, -1)
+}
+
+func (i *Inode) DenyWriteAccess() error {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 12]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+	if atomicDecUnlessPositive(&i.WriteCount) {
+		return nil
+	} else {
+		return syserror.ETXTBSY
+	}
+}
+
+func (i *Inode) AllowWriteAccess() {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack. Your proposal sounds good to me.

File: pkg/sentry/fs/inode.go [hunk 13]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+	atomic.AddInt32(&i.WriteCount, 1);
+}
+
+func (i *Inode) InodeIsOpenForWrite() bool {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 14]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+	if atomic.LoadInt32(&i.WriteCount) > 0 {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 15]
```diff
@@ -170,10 +258,21 @@ func (i *Inode) Lookup(ctx context.Context, name string) (*Dirent, error) {
...
+		err error
+	)
 	if i.overlay != nil {
-		return overlayCreate(ctx, i.overlay, d, name, flags, perm)
+		f, err = overlayCreate(ctx, i.overlay, d, name, flags, perm)
+	} else {
+		f, err = i.InodeOperations.Create(ctx, i, name, flags, perm)
 	}
-	return i.InodeOperations.Create(ctx, i, name, flags, perm)
+	if err == nil && flags.Write {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 16]
```diff
@@ -247,11 +346,25 @@ func (i *Inode) BoundEndpoint(path string) transport.BoundEndpoint {
 
 // GetFile calls i.InodeOperations.GetFile with the given arguments.
 func (i *Inode) GetFile(ctx context.Context, d *Dirent, flags FileFlags) (*File, error) {
+	var (
+		f   *File
+		err error
+	)
+	if flags.Write {
+		if err = d.Inode.GetWriteAccess(); err != nil {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 17]
```diff
@@ -170,10 +258,21 @@ func (i *Inode) Lookup(ctx context.Context, name string) (*Dirent, error) {
...
+	)
 	if i.overlay != nil {
-		return overlayCreate(ctx, i.overlay, d, name, flags, perm)
+		f, err = overlayCreate(ctx, i.overlay, d, name, flags, perm)
+	} else {
+		f, err = i.InodeOperations.Create(ctx, i, name, flags, perm)
 	}
-	return i.InodeOperations.Create(ctx, i, name, flags, perm)
+	if err == nil && flags.Write {
+		// No need to check return as it always success
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ack.

File: pkg/sentry/fs/inode.go [hunk 18]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+			break;
+		}
+	}
+	return true
+}
+
+// Refer to Linux/include/linux/fs.h for below helper functions on WriteCount.
+
+func (i *Inode) GetWriteAccess() error {
+	if !IsRegular(i.StableAttr) {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: My original purpose is to avoid such checks everywhere it's called. Do you have a specific reason to move it at the call site?

File: pkg/sentry/fs/inode.go [hunk 19]
```diff
@@ -63,6 +64,93 @@ type Inode struct {
...
+
+	if atomicIncUnlessNegative(&i.WriteCount) {
+		return nil
+	} else {
+		return syserror.ETXTBSY
+	}
+}
+
+func (i *Inode) PutWriteAccess() {
+	if !IsRegular(i.StableAttr) {
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/inode.go]: Ditto.

File: pkg/sentry/fs/ramfs/dir.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/sentry/fs/ramfs/dir.go]: GetWriteAccess() and PutWriteAccess() are called at inode/file level, not in a specified fs.

File: pkg/atomicbitops/atomic_bitops.go [hunk 20]
```diff
@@ -23,6 +23,10 @@
 // ordering. Using them to construct locks must employ proper memory barriers.
 package atomicbitops
 
+import (
```
[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/atomicbitops/atomic_bitops.go]: Oh, gVisor supports other platforms beyond arm64 and amd64? interesting. Will try 2nd option.

[the remark this answers is not in the document]
Comment (tanjianfeng) [reply pkg/atomicbitops/atomic_bitops.go]: Sorry, I choose an easier path (option 1). As I find that I did not have a ARM server to test for 2nd option (I can use a qemu binary translation if you insist on this option).

File: pkg/atomicbitops/atomic_bitops_common.go [hunk 21]
```diff
@@ -145,3 +145,45 @@ func DecUnlessOneInt32(addr *int32) bool {
 		}
 	}
 }
+
```
Comment (nlacasse) [inline pkg/atomicbitops/atomic_bitops_common.go]: Is this change still necessary?  Won't the methods in atomic_bitops_noarch get used?

Comment (tanjianfeng) [reply pkg/atomicbitops/atomic_bitops_common.go]: Oh, yes; strangely, I expect this leads to a compile failure, like duplicated symbol. Anyway, I'll drop this in *_common.go.

FYI, when I do rebase to latest code, it seems that, we need to adapt vfs2 to this:

`pkg/sentry/vfs/file_description_impl_util.go:260:23: cannot use fd (type *FileDescription) as type memmap.MappingIdentity in assignment:
        *FileDescription does not implement memmap.MappingIdentity (missing AllowWrite method)`

Comment (tanjianfeng) [reply pkg/atomicbitops/atomic_bitops_common.go]: Another thing, when I do rebase to latest code, it seems that, we need to adapt vfs2 to this:

`pkg/sentry/vfs/file_description_impl_util.go:260:23: cannot use fd (type *FileDescription) as type memmap.MappingIdentity in assignment:
        *FileDescription does not implement memmap.MappingIdentity (missing AllowWrite method)`

So far, I tend to add a TODO for that; will support it for vfs2 until vfs2 becomes stable.

File: pkg/atomicbitops/atomic_bitops_common.go [hunk 22]
```diff
@@ -129,19 +129,3 @@ func IncUnlessZeroInt32(addr *int32) bool {
 		}
 	}
 }
-
-// DecUnlessOneInt32 decrements the value stored at the given address and
-// returns true; unless the value stored in the pointer is 1, in which case it
-// is left unmodified and false is returned.
-func DecUnlessOneInt32(addr *int32) bool {
```
Comment (nlacasse) [inline pkg/atomicbitops/atomic_bitops_common.go]: This needs to stay

Comment (tanjianfeng) [reply pkg/atomicbitops/atomic_bitops_common.go]: Ah, my bad, thanks!

**Comment by tanjianfeng** [comment]
@nlacasse @nixprime I see some conflicts and test broken above. Would like to seek for your suggestion here. Shall we keep this patch for VFS, or we wait for VFS2?

**Comment by amscanne** [comment]
CC @nixprime Is this functionality already built in to VFS2, or should it be rebased?
